### PR TITLE
Fixes #709

### LINKED ISSUE #709: Torchlayer error when running on GPU
This is a continuation to the post here: https://discuss.pennylane.ai/t/pennylane-and-pytorch-running-on-gpu/457. 
I have two separate codes – one with classical machine learning (nothing to do with pennylane) and one with quantum ml (below). The only difference between the two is that the qml has additional pennylane code in the DQN class. Running the classical ml code runs without a problem on the GPU but when I run the qml code I get an error. Here is part of the code:

```
import gym
import math
import random
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
from collections import namedtuple
from itertools import count
from PIL import Image
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
import torchvision.transforms as T
from torch.nn.functional import relu, sigmoid
import pennylane as qml
import time

out_dim = 2  # output dimension of model
wires = 1  # this is the width of the quantum element
n_quantum_layers = 2  # this is the depth of the quantum element

def layer(inputs, w0, w1, w2, w3, w4, w5, w6, w7, w8, w9, w10):
    qml.templates.SqueezingEmbedding(inputs, wires=range(wires))
    qml.templates.CVNeuralNetLayers(w0, w1, w2, w3, w4, w5, w6, w7, w8, w9, w10,
                                    wires=range(wires))
    return [qml.expval(qml.X(wires=i)) for i in range(wires)]


class DQN(nn.Module):

    def __init__(self, img_height, img_width):
        super().__init__()
        self.flatten = nn.Flatten()
        self.fc1 = nn.Linear(in_features=img_height * img_width * 3, out_features=12)
        self.fc2 = nn.Linear(in_features=12, out_features=8)
       # self.fc3 = nn.Linear(in_features=10, out_features=8)
        self.clayer_in = torch.nn.Linear(in_features=8, out_features=wires)
        self.clayer_out = torch.nn.Linear(wires, out_dim)

        dev = qml.device('strawberryfields.fock', wires=wires, cutoff_dim=3)
        self.layer_qnode = qml.QNode(layer, dev)

        weights = qml.init.cvqnn_layers_all(n_quantum_layers, wires)
        weight_shapes = {"w{}".format(i): w.shape for i, w in enumerate(weights)}
        
        self.qlayer = qml.qnn.TorchLayer(self.layer_qnode, weight_shapes)

    def forward(self, t):
        t = self.flatten(t)
        t = self.fc1(t)
        t = self.fc2(t)
       # t = self.fc3(t)
        t = self.clayer_in(t)
        t = self.qlayer(t)
        t = self.clayer_out(t)
        t = t.sigmoid()
        return t

batch_size = 128
gamma = 0.999
eps_start = 1
eps_end = 0.01
eps_decay = 0.0005
target_update = 10
memory_size = 500000
lr_start = 0.01
lr_end = 0.00001
lr_decay = 0.00009
num_episodes = 1000 # run for more episodes for better results

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

em = CartPoleEnvManager(device)
strategy = EpsilonGreedyStrategy(eps_start, eps_end, eps_decay)
agent = Agent(strategy, em.num_actions_available(), device)
memory = ReplayMemory(memory_size)
#learning_rate = LearningRate(lr_start,lr_end,lr_decay)
#learn = lr(learning_rate)

policy_net = DQN(em.get_screen_height(), em.get_screen_width()).to(device)
target_net = DQN(em.get_screen_height(), em.get_screen_width()).to(device)
target_net.load_state_dict(policy_net.state_dict())
target_net.eval() #tells pytorch that target_net is only used for inference, not training
optimizer = optim.Adam(params=policy_net.parameters(), lr=0.01)

i = 0
episode_durations = []
for episode in range(num_episodes): #iterate over each episode
    program_starts = time.time()
    em.reset()
    state = em.get_state()
    
    for timestep in count():
        action = agent.select_action(state, policy_net)
        reward = em.take_action(action)
        next_state = em.get_state()
        memory.push(Experience(state, action, next_state, reward))
        state = next_state
        #i+=1
        #print(i)
        if memory.can_provide_sample(batch_size):
            scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=100, gamma=0.9)
            experiences = memory.sample(batch_size)
            states, actions, rewards, next_states = extract_tensors(experiences)
            
            current_q_values = QValues.get_current(policy_net, states, actions)
            next_q_values = QValues.get_next(target_net, next_states) #will get the max qvalues of the next state, q values of next state are used via next state
            target_q_values = (next_q_values * gamma) + rewards

            loss = F.mse_loss(current_q_values, target_q_values.unsqueeze(1))
            optimizer.zero_grad() # sets the gradiesnt of all weights n biases in policy_net to zero
            loss.backward() #computes gradient of loss with respect to all weights n biases in the policy net
            optimizer.step() # updates the weights n biases with the gradients that were computed form loss.backwards
            scheduler.step()
        if em.done:
            episode_durations.append(timestep)
            plot(episode_durations, 100)
            break
    if episode % target_update == 0:
        target_net.load_state_dict(policy_net.state_dict()) 
    now = time.time()
    print("Episode hat {0} Sekunden gedauert".format(now - program_starts))     
        
em.close()
```

And when running the code, the following error appears:

```
Traceback (most recent call last):
  File "qdqn.py", line 328, in <module>
    loss.backward() #computes gradient of loss with respect to all weights n biases in the policy net
  File "/home/ubuntu/anaconda3/envs/gymm/lib/python3.8/site-packages/torch/tensor.py", line 198, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/home/ubuntu/anaconda3/envs/gymm/lib/python3.8/site-packages/torch/autograd/__init__.py", line 98, in backward
    Variable._execution_engine.run_backward(
RuntimeError: Expected object of device type cuda but got device type cpu for argument #2 'mat2' in call to _th_mm
```

### PR DESCRIPTION
**Pull request by shukob**
**Context:**
In #709 it is reported that `TorchLayer` cannot property `backward` on CUDA. This PR fixes the issue.

**Description of the Change:**
Modified QNode parameter tensors to be converted into the input device.

**Benefits:**
We can use PyTorch layers on GPUs.

**Possible Drawbacks:**
I could not find any GPU related tests regarding torch, therefore we need to add some to ensure there're no drawbacks.

**Related GitHub Issues:**
#709

### CODE REVIEW

**Comment by josh146** [comment]
Thanks for the fix @shukob!

Perhaps this will also need to be fixed in `pennylane/tape/interfaces/torch.py`, so that CUDA works in [tape mode](https://pennylane.readthedocs.io/en/latest/code/qml_tape.html)?

**Review by trbromley — COMMENTED** [review]
Thanks @shukob! Looks great to me. Since we don't have tests, I checked using our local GPUs and confirmed that this fix works.

How about we add your [code snippet](https://github.com/PennyLaneAI/pennylane/issues/709#issuecomment-725954707) as a test to the torch qnn [tests](https://github.com/PennyLaneAI/pennylane/blob/master/tests/qnn/test_qnn_torch.py)? We could have the following to make sure we only test when GPUs are available:
```python
if not torch.cuda.is_available():
    pytest.skip("Cuda device not available")
```
Although our CI won't do these tests, at least we have a reliable way to run them locally. Could also sneak in an equivalent interface check there.

Otherwise, thanks for the PR! I agree with @josh146's suggestions too.

**Comment by shukob** [comment]
@josh146  Thank you for the comments.
I will include the `jacobian` device conversion for tape mode and changelog/contribution. It will not be able to be tested because `TorchLayer` seems to be not compatible with tape mode yet (as described in the document as "Tape-mode QNodes cannot be used with the qml.qnn module yet"). If I enable table mode in the snippet, the following error occurred.

```
AttributeError                            Traceback (most recent call last)

<ipython-input-1-1a5f6d72a6b2> in <module>
     15 weight_shapes = {"weights": (n_layers, n_qubits)}
     16 
---> 17 qlayer = qml.qnn.TorchLayer(circuit, weight_shapes)
     18 
     19 x = torch.rand((5, 10))

~/anaconda3/envs/***/lib/python3.7/site-packages/pennylane/qnn/torch.py in __init__(self, qnode, weight_shapes, init_method)
    208         super().__init__()
    209 
--> 210         self.sig = qnode.func.sig
    211 
    212         if self.input_arg not in self.sig:

AttributeError: 'function' object has no attribute 'sig'
```
@trbromley Thank you for the comments.
I will include the conditional test as you described.

**Comment by shukob** [comment]
It seems that at least `AmplitudeEmbedding` relies on NumPy functions to do padding on the initialization therefore it is not compatible with `torch.Tensor` inputs given in tape mode. Since `AmplitudeEmbedding` is generic and should not depend on torch, this should be handled before inputs are passed to qnode. Possible workarounds are
- We have conditional execution paths optimized for the respective input type. For example we can optimize `AmplitudeEmbedding` in a case where inputs are actually `torch.Tensor` on a `cuda` device.
- We always convert tensors onto CPU in tape mode.

**Comment by josh146** [comment]
>  It will not be able to be tested because TorchLayer seems to be not compatible with tape mode yet (as described in the document as "Tape-mode QNodes cannot be used with the qml.qnn module yet"). If I enable table mode in the snippet, the following error occurred.

@shukob this is a bit strange, since a recent PR updated `TorchLayer` to work in tape mode: https://github.com/PennyLaneAI/pennylane/pull/865.

Can you confirm if your branch is up to date with the latest master? If not, you may have found a bug/edge case in the implementation!

> This is a big leap in the direction, but if we have an abstract class to hold tensor's, we can delegate operations like padding to that class, so that tf uses tf.pad, torch uses torch.nn.functional.pad. Or we can have a functional proxy module that accepts numpy like functions and dispatch that to actual one.

Good catch! This is something we are aware of, and are attempting to go down this path at the moment 🙂 You can check out our progress here: #886

**Comment by shukob** [comment]
@josh146 You are right! I was executing the script on the v12.0 environment, sorry about that. It successfully executes on v13.0-dev. 

Thank you for the information about the agnostic tensor manipulation issue. It looks great! I'm looking for it to be implemented.
I am OK for the PR to be code reviewed.

File: tests/qnn/test_qnn_torch.py [hunk 1]
```diff
@@ -418,6 +418,34 @@ def test_gradients(self, get_circuit, n_qubits):
...
+        if not torch.cuda.is_available():
+            pytest.skip("Cuda device not available")
+        else:
+            try:
+                n_qubits = 4
+                dev = qml.device("default.qubit", wires=n_qubits)
+
+                @qml.qnode(dev, interface="torch")
+                def circuit(inputs, weights):
+                    qml.templates.AmplitudeEmbedding(inputs, wires=range(n_qubits), pad=0.0)
```
Comment (trbromley) [inline tests/qnn/test_qnn_torch.py:432]:
```suggestion
                    qml.templates.AngleEmbedding(inputs, wires=range(n_qubits))
```

File: tests/qnn/test_qnn_torch.py [hunk 2]
```diff
@@ -418,6 +418,34 @@ def test_gradients(self, get_circuit, n_qubits):
...
+
+                n_layers = 1
+                weight_shapes = {"weights": (n_layers, n_qubits)}
+
+                qlayer = qml.qnn.TorchLayer(circuit, weight_shapes)
+
+                x = torch.rand((5, 10))
+                norm = x.norm(p=2, dim=1, keepdim=True)
+                x_normalized = x.div(norm).to(torch.device("cuda"))
+                loss = torch.sum(qlayer(x_normalized)).squeeze()
```
Comment (trbromley) [inline tests/qnn/test_qnn_torch.py:444]:
```suggestion
                x = torch.rand((5, n_qubits), dtype=torch.float64)
                loss = torch.sum(qlayer(x)).squeeze()
```

File: tests/qnn/test_qnn_torch.py (hunk 1 above)
Comment (trbromley) [reply tests/qnn/test_qnn_torch.py:432]: Switched to AngleEmbedding as I was getting trouble with the numpy conversion in AmplitudeEmbedding

File: tests/qnn/test_qnn_torch.py (hunk 2 above)
Comment (trbromley) [reply tests/qnn/test_qnn_torch.py:444]: Note that the new `x` definition has the `dtype` specified - we need to cast to `torch.float64` to match the `dtype` of the qnode. This is more of a quick fix here, we should look at having the `qlayer` return type being dynamic, either set by the user upon instantiation or updated when fed an `x` of a certain type.

**Review by trbromley — APPROVED** [review]
Thanks @shukob! Approved, although I left some suggestions to get the test to work for me locally.

File: tests/qnn/test_qnn_torch.py (hunk 2 above)
Comment (trbromley) [reply tests/qnn/test_qnn_torch.py:444]: This could also be a side effect of my relatively old version of PyTorch.

**Comment by shukob** [comment]
@trbromley Thank you for the suggestions!

**Comment by shukob** [comment]
@trbromley It seems that `test_cuda_backward` does not actually run Cuda tensor. Sorry that I could not find the change. Can I post a small PR to fix this?

**Comment by josh146** [comment]
@shukob that would be great, thanks!

**Comment by shukob** [comment]
@josh146 Thanks!

**Comment by shukob** [comment]
Hi, I have found a performance degradation caused by this PR. Sorry about this. The problem is unnecessary implicit dtype conversion caused by the following lines.
```
                w = self.qnode_weights[arg].to(x)
```

```
        kwargs = {**{self.input_arg: x}, **{arg: weight.to(x) for arg, weight in self.qnode_weights.items()}}
```

These should be,

```
                w = self.qnode_weights[arg].to(x.get_device())
```

```
        kwargs = {**{self.input_arg: x}, **{arg: weight.to(x.get_device()) for arg, weight in self.qnode_weights.items()}}
```

I will make a PR for this if it is OK. Sorry again for multiple PRs.

**Pull request merged**